### ISSUE TITLE
Display differ a lot between before and after import

### ISSUE THREAD

**Issue by CohenCyril** [coq/coq#14105]
#### Description of the problem

I am very surprised about the following behaviour: why would the chosen notation differ before and after import?
To me the display from inside the module makes sense, but then, upon import it does not any more.

```coq
Module stuff.
Definition x := 0.
Module buggy.
Notation x := x.
End buggy.
Check x. (* displayed `x` OK to me *)
End stuff.
Check stuff.x. (* displayed `stuff.x` OK to me *)
Import stuff.
Check x. (* displayed `buggy.x` NOT OK not me *)
```

#### Coq Version

Coq 8.8 to 8.13

**Comment by CohenCyril** [comment]
CC @gares

**Comment by gares** [comment]
CC @herbelin

**Comment by SkySkimmer** [comment]
Reminiscent of https://github.com/coq/coq/issues/4904
The code at https://github.com/coq/coq/blob/271445decd0fc1a37da3009f148f2e68c7168fe1/interp/syntax_def.ml#L49-L57 should probably be changed somehow.

**Comment by gares** [comment]
If only I could understand it ;-)

**Comment by gares** [comment]
> if not syndef.syndef_onlyparsing then 

Maybe this can be used as a workaround for `Super.*` @CohenCyril ?

**Comment by CohenCyril** [comment]
> > if not syndef.syndef_onlyparsing then
> 
> Maybe this can be used as a workaround for `Super.*` @CohenCyril ?

This is slightly off-topic, but is it in Coq elpi's API?

**Comment by silene** [comment]
The conditional feels wrong. Should it not be a disjunction rather than a conjunction?

**Comment by herbelin** [comment]
I don't know exactly what would be the preferred semantics (what about collecting the different use cases and see what is expected in each case?). 

In any case, the change for #4904 looks strange though (as @silene wondered too in the meantime). I believe the test should at least about `i=1` and not `i<>1`, as in:
```
 let open_syntax_constant i ((sp,kn),(_local,syndef)) =
   let pat = syndef.syndef_pattern in
   if not (is_alias_of_already_visible_name sp pat) then begin
    (* Activate notation for parsing with appropriate qualifier *)
     Nametab.push_syndef (Nametab.Exactly i) sp kn; 
     (* Activate notation for printing at one level of Import only *)
     if Int.equal i 1 && not syndef.syndef_onlyparsing then
       Notation.declare_uninterpretation ~also_in_cases_pattern:syndef.syndef_also_in_cases_pattern (Notation.SynDefRule kn) pat 
   end 
```
Also, it is not clear to me how much it matters that the notation is `x` rather than say `y` in the report.

**Comment by CohenCyril** [comment]
> Also, it is not clear to me how much it matters that the notation is `x` rather than say `y` in the report.

```coq
Module stuff.
Definition x := 0.
Module buggy.
Notation y := x.
End buggy.
Check x. (* displayed `x` OK *)
End stuff.
Check stuff.x. (* displayed `stuff.x` OK *)
Import stuff.
Check x. (* displayed `buggy.y` NOT OK*)
```

**Comment by herbelin** [comment]
So, it is the same behavior, right?

**Comment by silene** [comment]
> I believe the test should at least about `i=1` and not `i<>1`

That indeed makes more sense.

There is still something that seems strange to me. While it makes sense to not register a syntactic definition if its symbol is already available, that does not mean that it will stay that way. If the symbol later gets shadowed, then the syntactic definition will not be available either. Am I missing something? (Perhaps this cannot happen?)

**Comment by herbelin** [comment]
> There is still something that seems strange to me. While it makes sense to not register a syntactic definition if its symbol is already available, that does not mean that it will stay that way. If the symbol later gets shadowed, then the syntactic definition will not be available either. Am I missing something? (Perhaps this cannot happen?)

Yes, I wondered about this too. This would probably be better to do this test at printing time rather than at import time, especially because `shortest_qualid_of_global` is rather context-dependent. I don't remember the use case I had in mind in 8659ff43db.

**Comment by herbelin** [comment]
For a comparison, `Zarith` is using the following pattern and (apparently) expects that the notation is exported:
```
Axiom Zpos : nat.
Module BinInt.
  Module Z.
  Notation pos := Zpos.
  End Z.
End BinInt.
Import BinInt.
Check Z.pos. (* Z.pos expected *)
```
which would be the same as the example with `y`:
```
Module stuff.
Definition x := 0.
Module buggy.
Notation y := x.
End buggy.
End stuff.
Import stuff.
Check buggy.y (* What to expect? *)
```
In the original example, we could still expect `x` even if typing `Check buggy.x`, on the argument that `buggy.x` is not a better form than `x`.

@cyrilcohen, what would be your own analysis?

**Comment by CohenCyril** [comment]
My own analysis is that there is currently a confusion between two features: namespacing (defining a new notation within a given namespace) on one hand and using namespace side-effects (choosing a preferred notation to display) on the other hand. This is linked to the current behavior of `Import` which has no granularity and does two jobs at the same time for notations in particular: 1. it shortens name and 2. it changes the display priority for a given term.

**Comment by herbelin** [comment]
Yes, import should be made more customizable.

I wonder whether I propose a PR or not and if yes with which semantics. Somehow, I'm tempted to continue living with the current behavior.

**Comment by silene** [comment]
I would not consider `ZArith` to be a good reason for keeping the current behavior. I find it quite annoying that Coq systematically replace the constructors `Zpos` and `Zneg` by the notations `Z.pos` and `Z.neg` in practice. This visual noise does not bring any advantage in my opinion.

**Comment by herbelin** [comment]
From the point of view of the designer of this notation, this was most probably not visual noise but instead a way to adopt the expected notation `Z.pos` and `Z.neg` without changing the historical names of the constructors.

This being said, I don't know what to conclude...

**Comment by proux01** [comment]
Fixed by https://github.com/rocq-prover/rocq/pull/20816
